OIDC doc update #475 (#480)
rucio/documentation: docs/operator/installing_server.md

### User Authentication Client
- **Grant Type**: `authorization_code`
- **Scopes**: Minimum `openid` and `profile`
- **Audience**: Recommended: `rucio`
- **Redirect URIs**: `https://<rucio_authorization_host>/auth/oidc_code`

#### Creating `idpsecrets.json`

For a single VO instance, create a file `idpsecrets.json` with the following content:

```json
{
  "def": {
      "user_auth_client": [
          {
              "issuer": "<issuer>",
              "client_id": "mock-client-id",
              "client_secret": "secret",
              "redirect_uris": "https://rucio/auth/oidc_code"
          }
      ]
  }
}
```
And please make sure you specify the path to this file either via env var `IDP_SECRETS_FILE` or in rucio.cfg as
```cfg
[oidc]
idpsecrets = /path/to/your/idpsecrets.json
```

#### Requiring extra Scopes
If you want to add extra scope of ID token for authentication you can add it as required in server.

Scopes can be extended if needed using the following configuration and corresponding claims existence for the scopes
```cfg
[oidc]
id_token_extra_scopes= email, extra_scope
id_token_extra_claims= email, claim_of_extra_scope
```

> Only claim existence is check for now. If you need functionality for exact claim's value match please contact us we can add that feature too.

Reviewer: Expanding this to add the feature request link to clarify where they should reach out

```suggestion
Only claim existence is checked for now. Please open a [feature request](https://github.com/rucio/rucio/issues/new?template=feature.yaml) if you would be interested in having this check expanded to take into account the claim value as well.
```

> Rucio will exchange the authorization code using user_auth_client for an ID token and an access token. If additional scopes are required for the access token, configure them as follows:

Reviewer: ```suggestion
Rucio will exchange the authorization code using `user_auth_client` for an ID token and an access token. If additional scopes are required for the access token, configure them as follows:
```

[oidc]
extra_access_token_scope= extra_access_scopes

Ensure all required scopes are included as needed.

#### Adding user identity

> To add user oidc identity you ned to get `SUB`, Subject Identifier, of the user in the IDP. Then run the command:

Reviewer: ```suggestion
To add user oidc identity you need to get the `SUB` (Subject Identifier) of the user in the IDP. Then run the command:
```

```shell
rucio account identity add --account jdoe --type OIDC --id 'SUB=<sub> ISS=<issuer>'
```
or legacy command
```shell
rucio-admin identity add --account jdoe --type OIDC --id 'SUB=<sub> ISS=<issuer>'
```
Example for IAM instance
```shell
rucio account identity add --account jdoe --type OIDC --id 'SUB=3ed4fg-6ff2-4097-ad3b-953e11bb52b8 ISS=https://panda-iam-doma.cern.ch/'
```

#### Multiple IDP

> To configure multiple IDPs your `idpsecrets.json` looks like:

Reviewer: ```suggestion
To configure multiple IDPs, your `idpsecrets.json` should look like:
```

```json
{
  "def": {
      "user_auth_client": [
          {
              "issuer": "<issuer>",
              "client_id": "mock-client-id",
              "client_secret": "secret",
              "redirect_uris": "https://rucio/auth/oidc_code",
              "issuer_nickname": "issuer1"
          },
          {
              "issuer": "<issuer>",
              "client_id": "mock-client-id2",
              "client_secret": "secret2",
              "redirect_uris": "https://rucio/auth/oidc_code",
              "issuer_nickname": "issuer2"
          }
      ]
  }
}
```

> Notice now we *must* add `issuer_nickname` field to each idp. This is used by client to reference which issuer it is authenticating to.

Reviewer: ```suggestion
Notice now we *must* add an `issuer_nickname` field to each IDP. This is used by the client to reference which issuer it is authenticating to.
```

### Transfer

> To use tokens for transfers you need to register another client lets call it `client_credential_client` which needs following.

Reviewer: ```suggestion
To use tokens for transfers, you need to register another client (in this example, `client_credential_client`), configured as such:
```

- **Grant Type**: `client_credentials`
- **Scopes**: `fts`, `storage.read:/<path>` and `storage.modify:/<path>`
- **Audience**: `<fts_hostname>` and `<RSE_hostname>`

If you want to allow FTS to refresh storage token, allow refresh token to be returned too i.e. add `offline_access` in the scopes.

> #### **Configuration File (idpsecrets.json) Format (Single VO Syntax)**

Reviewer: ```suggestion
#### **Configuration File (`idpsecrets.json`) Format (Single VO Syntax)**
```

```json
{
  "def": {
    "user_auth_client": [
          {
              "issuer": "<issuer>",
              "client_id": "mock-client-id",
              "client_secret": "secret",
              "redirect_uris": "https://rucio/auth/oidc_code"
          }
      ],
      "client_credential_client": {
          "client_id": "<your_client_id>",
          "client_secret": "<your_client_secret>",
          "issuer": "<issuer>"
      }
  }
}
```

#### Configuration for RSE

> For RSEs which uses token, RSEs must have an attribute set as follows:

Reviewer: ```suggestion
For RSEs which use tokens, RSEs must have an attribute set as follows:
```

```
oidc_support: True
```

> Lets say your RSE has protocol with prefix `/path/myexp/mypath` , if you want to use this as scope you needs to have `client_credential_client`

Reviewer: ```suggestion
Let's say your RSE has protocol with prefix `/path/myexp/mypath` , if you want to use this as scope you need to have `client_credential_client`
```

> scope `storage:read:/path/myexp/mypath` and `storage:modify:/path/myexp/mypath`.

Reviewer: ```suggestion
scopes `storage:read:/path/myexp/mypath` and `storage:modify:/path/myexp/mypath`.
```
